Hugging Face Tokenizer and Datasets

nprisbrey opened the pull request:
This pull request overhauls the way we load data and create DataLoaders. Due to the significance of the change and conflicting namespaces, I've replaced the datasets.py file with a new load_data.py file.

The old code in datasets.py was hard-coded to only accept one version of wikitext. The new data loading code allows the user to choose any [text dataset on Hugging Face](https://huggingface.co/datasets?task_categories=task_categories:text-generation) and -- through the `--dataset-name`, `--dataset-subset`, and `--dataset-feature` command-line arguments -- use the dataset to train the model.

As for the tokenizer, I've implemented the BPE tokenizer quite similarly to [the Hugging Face tutorial](https://huggingface.co/learn/nlp-course/chapter6/8?fw=pt#building-a-bpe-tokenizer-from-scratch). Let it be noted that there are 3 other tokenizers supported by Hugging Face that could be implemented in the future (if desired).

I've chosen to only train the tokenizer on the training dataset, thereby maintaining the integrity of the validation and testing datasets as data that hasn't been seen before.

While writing this code, I explored appending an end-of-sequence token to sequences but it became messy. As stated in the RetNet paper, however, I have included a beginning-of-sequence token for each sequence. Due to these changes, I see this PR as resolving Issue https://github.com/DRAGNLabs/301r_retnet/issues/6.

I've also updated the scripts to train the models to include all the possible arguments, be alphabetized, and include numbers closer to the scale that we'll want to implement in a data run.

All formatting, comments, and doc strings have been updated, though I haven't touched the README.md. It is my feeling that the README needs a large overhaul and would be worth completing in a separate Pull Request.

nprisbrey committed:
First Attempt At Hugging Face Dataset And Tokenizer Implementation

nprisbrey committed:
Ignore Datasets data/ Directory

nprisbrey committed:
Implement Hugging Face Tokenizer During Text Generation

nprisbrey committed:
Update Setting Random Seed To Affect More Libraries

nprisbrey committed:
Slight Tweaks + Remove Support For Non-BPE Tokenizers

Remove support for non-BPE tokenizers. Could be implemented at another
time if desired. Also remove the end-of-sequence tokens due to
problems with data processing and this token not being specified in
the original RetNet paper. Fix bug where tokenizer was only training
on the word "text" instead of the actual dataset.

nprisbrey committed:
Save Tokenizer Along With Model At End Of Training

nprisbrey committed:
Add Arguments To Use Other Hugging Face Datasets

nprisbrey committed:
Support All Datasets With Any Splits

Change functionality to manually create train, validation, and
test datasets instead of only accepting datasets that come with
the three splits premade.

nprisbrey committed:
Update Example Scripts

Include all possible arguments as examples of possible inputs. Reorder
arguments to be in alphabetical order. Update values to be closer to
what would be used in a data run.

nprisbrey committed:
Make Code Format Consistent

nprisbrey committed:
Update load_data.py Documentation And Comments

nprisbrey committed:
Move Dataset Splits To After Filtering

The dataset splits need to be after filtering instead of before.
Splitting before filtering could lead to inequal splits after
removing instances.

nprisbrey committed:
Shuffle Data Before Dataset Splits

KimballNJardine commented on load_data.py:
```diff
@@ -0,0 +1,115 @@
+        Tuple with the format: (Training DataLoader, Validation DataLoader,
+        Testing DataLoader, Tokenizer object).
+    """
+    assert ["test", "train", "validation"] == sorted(get_dataset_split_names(path=dataset_name, config_name=dataset_config)), \
```
Is this meant to be different order between the assert statement and the message?
assert --> test, train, validation
print   --> train, validation, test

nprisbrey replied:
This code was not ready for review yet, hence the draft PR. The code has now been updated and I'm changing this to a "real" PR, meaning it's now ready for your review. :slightly_smiling_face:

KimballNJardine commented on load_data.py:
```diff
@@ -0,0 +1,115 @@
+                             special_tokens=["<pad>", "<bos>", "<eos>", "<unk>"],
+                             max_token_length=max_token_len)
+    elif tokenizer_name == "Unigram":
+        tokenizer = Tokenizer(Unigram())
```
Does Unigram need unk_koten="<unk>" added here or left out?

KimballNJardine commented on load_data.py:
```diff
@@ -0,0 +1,115 @@
+        tokenizer = Tokenizer(BPE(unk_token="<unk>"))
+        trainer = BpeTrainer(vocab_size=vocab_size,
+                             show_progress=True,
+                             special_tokens=["<pad>", "<bos>", "<eos>", "<unk>"],
```
I think it would be better to have a special_tokens variable as a list above called something like special_tokens, and simply passing it to minimize code duplication and allow for easy alterations.

KimballNJardine commented on load_data.py:
```diff
@@ -0,0 +1,115 @@
+        tokenizer = Tokenizer(Unigram())
+        trainer = UnigramTrainer(vocab_size=vocab_size,
+                                 show_progress=True,
+                                 special_tokens=["<pad>", "<bos>", "<eos>", "<unk>"],
```
See above for special_tokens comment

KimballNJardine commented on load_data.py:
```diff
@@ -0,0 +1,115 @@
+        tokenizer = Tokenizer(WordLevel(unk_token="<unk>"))
+        trainer = WordLevelTrainer(vocab_size=vocab_size,
+                                   show_progress=True,
+                                   special_tokens=["<pad>", "<bos>", "<eos>", "<unk>"])
```
See above for special tokens comment.

KimballNJardine commented on load_data.py:
```diff
@@ -0,0 +1,115 @@
+                                        max_input_chars_per_word=max_token_len))
+        trainer = WordPieceTrainer(vocab_size=vocab_size,
+                                   show_progress=True,
+                                   special_tokens=["<pad>", "<bos>", "<eos>", "<unk>"])
```
See above for special tokens comment.

KimballNJardine commented on load_data.py:
```diff
@@ -0,0 +1,115 @@
+                          data_dir: str,
+                          dataset_config: str=None,
+                          max_token_len: int=20) -> tuple[DataLoader, DataLoader, DataLoader, Tokenizer]:
+    """ Loads the WikiText2 dataset and returns DataLoaders.
```
This is not necessarily a WikiText2 dataset. Could be another kind.

KimballNJardine commented on train_model.py:
```diff
@@ -228,6 +228,9 @@ def get_params(self) -> dict:
             help="Random seed to use, allowing more reproducible results.")
     parser.add_argument("-s", "--seq-len", type=int, default=512,
             help="Sequence length (context window size).")
+    parser.add_argument("-t", "--tokenizer", type=str, default="BPE",
```
If possible, update the shell files for testing to have -t BPE in them for clarity.

KimballNJardine commented on train_model.py:
```diff
@@ -325,10 +328,15 @@ def get_params(self) -> dict:
     print(f"-log(1 / {args.vocab_size}) = {-torch.log(torch.tensor(1 / args.vocab_size))}")
 
     # Get DataLoaders
```
If a TODO can be left noting that dataset_name and dataset_config will need to be improved to not just default to wikitext for a future pull request that would be good. However, that is not required.

nprisbrey replied:
I agree that for a serious data run in the future, we'll need to use a larger dataset (such as the c4 dataset we talked about earlier). I feel that the default should be left to a small dataset such as wikitext, however, to allow new people to the repository to try out the code, train models, and get results without knowing exactly what they're doing. If we default to a larger dataset like c4, I worry that we start limiting those that can explore our code base to those with supercomputers/fast download speeds/large amounts of memory. I would personally push for leaving a simpler dataset like wikitext here and encouraging others to try larger datasets in the README for new people arriving at the repository that want to explore further. What do you think?

KimballNJardine commented on train_model.py:
```diff
@@ -350,11 +358,11 @@ def get_params(self) -> dict:
         train_total_samples = 0
-        for batch_idx, (inputs, targets) in enumerate(tqdm(train_loader,
-                                                           desc="Train")):
+        for batch_idx, batch_seqs in enumerate(tqdm(train_loader,
```
maybe call this train_batch_seqs to follow the pattern of val_batch_seqs below. However, this is not required in my mind. Can be discussed.

KimballNJardine commented on train_model.py:
```diff
@@ -439,10 +446,10 @@ def get_params(self) -> dict:
     total_samples = 0
     with torch.inference_mode():
-        for inputs, targets in tqdm(test_loader, desc="Test"):
+        for batch_seqs in tqdm(test_loader, desc="Test"):
```
Maybe call this test_batch_seqs to follow pattern of val_batch_seqs above. However, not required. Can be discussed.

nprisbrey replied:
Good call. This has been addressed in [eb4bffc](https://github.com/DRAGNLabs/301r_retnet/pull/15/commits/eb4bffc2bbb4cf79f273b1f704ab63f37663e126).

nprisbrey commented:
@KimballNJardine, you recognize that the 5 comments above are for outdated versions of the code, correct? The referenced code has already been updated or removed in later commits.

KimballNJardine commented on scripts/transformer.sh:
```diff
@@ -1,15 +1,23 @@
     --rand-seed 42 \
+    --seq-len 128 \
+    --splits 0.7 0.2 0.1 \
+    --val-freq 3 \
```
If you can add 
-t BPT \
that would be good. It will make no difference but make it explicit for testing.

KimballNJardine commented on train_model.py:
```diff
@@ -460,34 +497,39 @@ def get_params(self) -> dict:
     print(f"Saved final weights as {weight_filename}")
 
+    # Save BPE tokenizer used
+    tokenizer.save_pretrained(save_directory=save_folder, filename_prefix="BPE")
```
Please move this to the beginning before training begins. The vocabulary is needed, especially if the training crashes.

nprisbrey replied:
Well said! This has been addressed in [54a977d](https://github.com/DRAGNLabs/301r_retnet/pull/15/commits/54a977dbb687efbeee52151b3e13cfbc3c2748cc).

KimballNJardine reviewed (changes requested):
I realized that I was reviewing an outdated pull request, so most of the other comments I have made are outdated. Please move saving the tokenizer information to the beginning of the run, not the end.

KimballNJardine commented on train_model.py:
```diff
@@ -439,10 +476,10 @@ def get_params(self) -> dict:
     total_samples = 0
     with torch.inference_mode():
-        for inputs, targets in tqdm(test_loader, desc="Test"):
+        for batch_seqs in tqdm(test_loader, desc="Test"):
```
rename this test_batch_seqs to match val_batch_seqs, as seen above for consistency.

KimballNJardine commented on train_model.py:
```diff
@@ -350,11 +378,12 @@ def get_params(self) -> dict:
         train_total_samples = 0
-        for batch_idx, (inputs, targets) in enumerate(tqdm(train_loader,
-                                                           desc="Train")):
+        for batch_idx, batch_seqs in enumerate(tqdm(
```
Maybe call this train_batch_seqs to follow pattern of val_batch_seqs below. However, not required. Can be discussed.

nprisbrey committed:
Standardize "*batch_seqs" Variables In train_model.py

nprisbrey committed:
Save Trained Tokenizer Immediately After Training

Moved saving the trained tokenizer to immediately after training
instead of after LLM model training. This prevents data loss if
training needs to end early or an error occurs later in the script.

nprisbrey committed:
Remove Extra Shuffle Of Dataset In Preprocessing

Due to an oversight in cc6928d, datasets were being shuffled twice.

KimballNJardine commented on load_data.py:
```diff
@@ -0,0 +1,159 @@
+    entire_dataset = entire_dataset.filter(filter_fun)
+
+    # Split into training, validation, and testing datasets
+    train_testvalid = entire_dataset.train_test_split(
```
My understanding of this code is you split two times using the train_test_split() function.
train = train_testvalid["train"]
validation = test_valid["train"]
test = test_valid["test"]

However, implicit in the ordering of the variable name train_testvalid is that train is under ["train"], and testvalid is under ["test"].
My issue is that test_valid would implicitly suggest that test is under ["train"], and validation is under ["test"]. The code says otherwise.
**If you could rename the variables train_testvalid and test_valid to be train_validtest and valid_test, respectively, that would make the code a bit clearer to read.** Furthermore, it would follow the order the comment above states (#split into training, validation, and testing datasets).

nprisbrey replied:
Good catch! Even reviewing this myself after you pointed it out made me confused for a minute. I like your idea of renaming, thereby attempting to keep things clearer. I've made this change in [af42f88](https://github.com/DRAGNLabs/301r_retnet/pull/15/commits/af42f88d521e8896ce9b9803b8b641a256664662).

KimballNJardine reviewed (changes requested):
I have only a minor request involving variable names. If you have an issue with my request, please let me know. Otherwise, it looks good.

nprisbrey committed:
Improve Confusing Variable Names When Splitting Dataset

KimballNJardine reviewed (approved):
I see no issues. Approved.
